core/sbt-cloudflow/src/main/scala/cloudflow/sbt/CloudflowBasePlugin.scala

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (C) 2016-2020 Lightbend Inc. <https://www.lightbend.com>
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cloudflow.sbt
+
+import java.nio.charset.StandardCharsets._
+import java.util.Base64
+import java.util.zip.Deflater
+import java.io.ByteArrayOutputStream
+
+import com.typesafe.config._
+
+import sbt._
+import sbt.Keys._
+import sbtdocker.DockerKeys._
+import com.typesafe.sbt.packager.archetypes._
+import spray.json._
+
+import cloudflow.sbt.CloudflowKeys._
+import cloudflow.blueprint.StreamletDescriptorFormat._
+import cloudflow.blueprint.StreamletDescriptor
+
+/**
+ * Base class for all Cloudflow runtime plugins for multi-image use case. Contains some
+ * methods which are reused across the runtime plugins, `CloudflowFlinkPlugin`,
+ * `CloudflowAkkaPlugin` and `CloudflowSparkPlugin`.
+ */
+abstract class CloudflowBasePlugin extends AutoPlugin {
+  final val AppHome = "${app_home}"
+  final val AppTargetDir: String = "/app"
+  final val appTargetSubdir: String ⇒ String = dir ⇒ s"$AppTargetDir/$dir"
+  final val AppJarsDir: String = "app-jars"
+  final val DepJarsDir: String = "dep-jars"
+  final val OptAppDir = "/opt/cloudflow/"
+  final val scalaVersion = "2.12"
+
+  // NOTE !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+  // The UID and GID of the `jboss` user is used in different parts of Cloudflow
+  // If you change this, you have to make sure that all references to this value are changed
+  // - fsGroups on streamlet pods uses the GID to make volumes readable
+  val UserInImage = "185" // default non-root user in the spark image
+  val userAsOwner: String ⇒ String = usr ⇒ s"$usr:cloudflow"
+  val StreamletDescriptorsLabelName = "com.lightbend.cloudflow.streamlet-descriptors"
+
+  override def requires = CommonSettingsAndTasksPlugin && StreamletScannerPlugin &&
+    JavaAppPackaging && sbtdocker.DockerPlugin
+
+  private[sbt] val verifyDockerRegistry = Def.task {
+    cloudflowDockerRegistry.value.getOrElse(throw DockerRegistryNotSet)
+  }
+
+  private[sbt] val checkUncommittedChanges = Def.task {
+    val log = streams.value.log
+    if (cloudflowBuildNumber.value.hasUncommittedChanges) {
+      log.warn(s"You have uncommitted changes in ${thisProjectRef.value.project}. Please commit all changes before publishing to guarantee a repeatable and traceable build.")
+    }
+  }
+
+  private[sbt] def makeStreamletDescriptorsLabelValue(streamletDescriptorsJson: JsValue) = {
```

Comment:
This is more of a helper-static method to be used from a companion object. Same for the other ones that dont return a Task.

Reply:
Initially instead of an abstract base class `CloudflowBasePlugin` I made it an object `PluginUtils` and did an `import PluginUtils._` form the individual plugins. Then I figured out that we can reuse more if we make it a parent plugin e.g. `requires` etc. Hence the current shape. I think extracting out stuff into another companion object will be a bit of an over-engineering. We can view this as a base class that offers some _common_ behaviors that can be reused in derived classes. And it's not necessary that each of them will have to return a `Task` ..